File: 2289.steps-to-make-array-non-decreasing.java
#
# @lc app=leetcode id=2289 lang=java
#
# [2289] Steps to Make Array Non-decreasing
#
# @lc code=start
class Solution {
    public int totalSteps(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        int res = 0;
        java.util.Stack<Integer> stack = new java.util.Stack<>();
        for (int i = 0; i < n; ++i) {
            int maxSteps = 0;
            while (!stack.isEmpty() && nums[i] >= nums[stack.peek()]) {
                maxSteps = Math.max(maxSteps, dp[stack.pop()]);
            }
            if (!stack.isEmpty()) {
                dp[i] = maxSteps + 1;
                res = Math.max(res, dp[i]);
            }
            stack.push(i);
        }
        return res;
    }
}
# @lc code=end

File: 2289.steps-to-make-array-non-decreasing.go
#
# @lc app=leetcode id=2289 lang=golang
#
# [2289] Steps to Make Array Non-decreasing
#
# @lc code=start
func totalSteps(nums []int) int {
    n := len(nums)
    stack := make([][2]int, 0) // Each entry: [value, steps]
    res := 0
    for i := n - 1; i >= 0; i-- {
        steps := 0
        for len(stack) > 0 && nums[i] > stack[len(stack)-1][0] {
            steps = max(steps+1, stack[len(stack)-1][1])
            stack = stack[:len(stack)-1]
        }
        res = max(res, steps)
        stack = append(stack, [2]int{nums[i], steps})
    }
    return res
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
# @lc code=end